src/object.d

```diff
@@ -1718,6 +1745,21 @@ class Throwable : Object
      */
     @system @nogc final pure nothrow ref uint refcount() return scope { return _refcount; }
 
+    /**
+     * Loop over the chain of Throwables.
+     */
+    int opApply(scope int delegate(ref Throwable) dg)
```

Comment:
Not sure if this isn't operator abuse. Maybe it would be more natural if an iteratable is returned from a `chain()` property?

Why did you mix this change into this PR? Why is the test for it in https://github.com/dlang/dmd/pull/7832?

Reply:
> Not sure if this isn't operator abuse. Maybe it would be more natural if an iteratable is returned from a chain() property?

That's what opApply is for, I don't know why it would abuse. Iterables are not normally done in D.

> Why did you mix this change into this PR?

Because of the complaints I received that the PR was incomplete. Nobody was interested in doing this in stages.

> Why is the test for it in dlang/dmd#7832?

Because that's where the tests are for the chained exceptions.

Reply:
The preferred iteration mode in D is by means of ranges. Uses of `opApply` should provide rationale on why internal iteration is preferred or necessary.

The delegate signature should take a `Throwable`, not a `ref Throwable`. Also it seems `scope Throwable` would be better because we don't want to escape `Throwable` objects during iteration.

All in all this seems to need careful planning.

Reply:
> That's what opApply is for, I don't know why it would abuse.

Without knowing all the (implementation) details of Throwable, I would not expect to be able to iterate over an exception.

> Because of the complaints I received that the PR was incomplete.

opApply is unrelated to ref counting. AFAICT reference counting is not used so far, so I actually didn't scrutinize the modified reference counting in this PR because I expect it to change anyway.

> Because that's where the tests are for the chained exceptions.

Iterating with foreach over the exception chain is syntax sugar irrelevant to the compiler implementation. I don't like all the unittests in object.d, so I sympathize with putting them elsewhere, but having them in another repo is also a complication.

Reply:
> I would not expect to be able to iterate over an exception.

Yet existing code *does*.

> having them in another repo is also a complication.

Yes, but that's beyond the scope of this PR.